pkg/controller/framework/controller.go

```diff
@@ -119,13 +154,8 @@ func (c *Controller) Requeue(obj interface{}) error {
 func (c *Controller) processLoop() {
 	for {
 		obj := c.config.Queue.Pop()
-		err := c.config.Process(obj)
-		if err != nil {
-			if c.config.RetryOnError {
-				// This is the safe way to re-enqueue.
-				c.config.Queue.AddIfNotPresent(obj)
-			}
-		}
+		// TODO: Dispatch it correctly.
+		c.workQueue[0] <- obj
```

Comment:
So now that I look at this, I don't think you can shard here at all without changing the semantics of this framework. By sharding here, you'll no longer be processing things in order. Users of this might really depend on this property.